### ISSUE TITLE
Payment modul

### ISSUE THREAD

**Issue by maltitco** [awesto/django-shop#579]
I have a question about creating a payment module for one of the operators.
The operator I want to integrate requires that POST data to the transaction log after it returns a token for further authorization.
One of these data is OrderId, which can be sent only once to the payment system.
After this authorization and status check is carried out by the token given when the payment is registered.

Should I implement the payment as it is with the "Pay in advance" method?
Where an order is created with the possibility of later payment.

Or is it possible to give the orderID in advance to the operator and go to his site to pay for the order?

I have tracked how it is in the case of paypal payment and there the orderID is created only after returning from paypal with information on the correct payment of the order.

**Comment by jrief** [comment]
Well noticed!

Indeed, there is no way to create an Order object, before proceeding to the PSP.

Therefore I moved the code to populate the Order out of ``create_from_cart()``.
Please read the release notes on this branch:
https://github.com/jrief/django-shop/tree/drafts/create-order-before-populating

and try using it in your environment.

**Comment by jrief** [comment]
@maltitco did you already try that patch?

**Comment by maltitco** [comment]
Hi Jacob,
Unfortunately I did not have time to check it out.
I will try to check it next week.

**Comment by maltitco** [comment]
Hi Jacob,
I'm trying to implement slowly but I'm not really going. I'm based on the PayPal module.
How can I pass the order number to `def get_payment_request`?

**Comment by jrief** [comment]
why don't you use the PayPal module, from Awesto?

**Comment by maltitco** [comment]
You misunderstood me, I am creating a module for a local PSP but I see how it is resolved in Awesto PayPal module.

As I wrote earlier, I have to give more information from ordering to register a PSP transfer to receive a token to further identify the order in their system.

**Comment by jrief** [comment]
If you click on `Purchase Now` the REST endpoint delegates this event to the PSP's `get_payment_request` method. It is here where you can create the Order object, it remains in state `new`. It's here where you can `get_or_assign_number()` – the order number.

Then, when the PSP accepts the payment and redirects the client to the `return_view` (success handler) of your Payment module, add the payment transaction id and save the order. Add a transition, so that the order object has a status of `paid_with_foo`.

Remember that the `get_payment_request` can return executable Javascript, which is `eval()`ed by the client.

**Comment by maltitco** [comment]
OK, I have one more question.
PSP allows to download token only once for one unique order number.
Everything is ok if the payment goes well, but when the payment is wrong and the customer tries to re-order, the django-shop gives him the same order number and the PSP rejects it because for that number the token was downloaded.

**Comment by jrief** [comment]
Sorry for the delay.
So what you say, is that the PSP declined the payment, btw. which PSP?
If the order has already been created, then we have two options.
1. Let the customer re-add the order's content to the cart. There is a special method for this: ``MyOrder.readd_to_cart``. After that he can proceed as normal.
2. There is no user interface yet, but I could add one to the order's detail view for declined payments, so that you can re-add the payment information. If that's a better option for you, I will check how to do it.

**Comment by maltitco** [comment]
Or perhaps it would be better to let the customer create an order as it does for "Pay in advance" and on the order confirmation view, add a button with the option to paying for order which is already redirected to PSP.
In this case, we have all the necessary data to fill out the form for the PSP.

It would be useful for the customer, after a failed payment to be able to choose another payment method. So for example, after rejection of a CreditCard payment, the customer returns to his order and can choose to pay by cash, bank transfer, etc.

Re-adding products to the cart will not be possible after running the warehouse status management (The quantity of the product has changed to 0 and no longer can be ordered.), what shall I ask about soon.

You ask about PSP that is eService, but I want to do integration for several Polish operators so that I can choose the one that will offer the best conditions.

**Comment by jrief** [comment]
The pull request for proposal is not merged yet. Since it breaks the API, I have to wait until 0.11 anyway. If you can provide an interface to your PSP on existing (but unpaid) orders, that's perfectly fine.

**Comment by maltitco** [comment]
So I will do, it will be a much more flexible solution. If you can add the ability to change the payment method after the unpaid order, this would be a useful feature.

I will create it to the current version and after release 0.11 adapt to the new one.

**Comment by maltitco** [comment]
I have a question about how to display different templates, depending on the selected type of payment method.

**Comment by jrief** [comment]
You have to add a template to your project named ``myshop/checkout/payment-method-form.html``. Extend it from its counterpart from django-SHOP. In ``{% block payment-provider-form %}`` add whatever you need. Use ``ng-show`` or ``ng-switch`` to dynamically show the payment form for the chosen PSP.

**Comment by maltitco** [comment]
Another question:
I'm trying to redirect to the PSP url with the data sent by POST using the requests library

        s = requests.Session()
        r = s.post(url, data=payload)
        js_expression = '$window.location.href="{}";'.format(r.url)
        return js_expression

The URL opens but there is no data posted by POST

**Comment by jrief** [comment]
Is ``url`` an endpoint of your PSP and you try to post ``payload`` to it?
In which function do you invoke this? May it be that you must wrap it into a JS promise?
Without knowledge of your PSP's API, it's difficult to say, what's wrong.

**Comment by maltitco** [comment]
Yes the url is an endpoint of PSP.
Return function running the Proceed to: PURCHASE_NOW - which one is redirected to the PSP page to the payment form.

if I change:
`r = s.post(url, data=payload)`
to:
`r = s.post(url, params=payload)`

Redirect to PSP payment page with form but transaction data is visible in URL addres.

**Comment by maltitco** [comment]
I have another question, this time about changing the order status by return url.
My operator, depending on the status of payment, redirects the customer to the different return address in the store. Is it possible to change payment status depending on your return address?

**Comment by jrief** [comment]
Disclaimer: without having tested this myself, but I think this approach should work:

1. Create your own OrderWorkflowMixin with some methods:
2. for each payment status your PSP provides, create a method decorated with `@transition(...)`.
3. create different Order Views for each possible status
4. each of these Order Views shall call one of the methods of your WorkflowMixin.

**Comment by maltitco** [comment]
@jrief Thank you for the tips, I will try to do it the way you described it.

**Comment by jrief** [comment]
On 0.11 this has been changed. So the workflow is
1. create Order (status = new)
2. assign Order number
3. Handle payment at PSP
4. If the payment fails the Order remains new, but the Cart is still filled
5. If the payment was successful, populate the order items from the cart

you can of course add another Order object, where the order is filled, even for failed payments. They then could be handled by another kind of payment.

Please create a new issue, if that's required by your workflow.
